osu.Game/Overlays/MusicController.cs

```diff
@@ -385,15 +387,27 @@ public bool AllowTrackAdjustments
         /// </remarks>
         public void ResetTrackAdjustments()
         {
+            // todo: we probably want a helper method rather than this.
             CurrentTrack.RemoveAllAdjustments(AdjustableProperty.Balance);
             CurrentTrack.RemoveAllAdjustments(AdjustableProperty.Frequency);
             CurrentTrack.RemoveAllAdjustments(AdjustableProperty.Tempo);
             CurrentTrack.RemoveAllAdjustments(AdjustableProperty.Volume);
 
-            if (allowTrackAdjustments)
+            modTrackAdjustments.RemoveAllAdjustments(AdjustableProperty.Balance);
+            modTrackAdjustments.RemoveAllAdjustments(AdjustableProperty.Frequency);
+            modTrackAdjustments.RemoveAllAdjustments(AdjustableProperty.Tempo);
+            modTrackAdjustments.RemoveAllAdjustments(AdjustableProperty.Volume);
+
+            var applicableToTrack = mods.Value.OfType<IApplicableToTrack>();
+
+            if (!allowTrackAdjustments || !applicableToTrack.Any())
+                CurrentTrack.UnbindAdjustments(modTrackAdjustments);
```

Comment:
I see that this has changed somewhat from the original diff you posted (and is now more correct), but my gripe here is thus: how is this any safer than the previous code? Any mod is still free to hold references to `modTrackAdjustments` and mess with them after it's not supposed to, because it's still a single shared reference.

If we're so insistent on using the adjustment layer, then why doesn't this method just create a *new* `AudioAdjustments` every time it's called, discard the previous one, and expose the new temporary one to the mods? Which is to say:

```csharp
private AudioAdjustments modAdjustments;

public void ResetTrackAdjustments()
{
    // remove everything from CurrentTrack as in the original

    CurrentTrack.UnbindAdjustments(modAdjustments);

    // statement below equivalent to removing all adjustments as done in your version
    CurrentTrack.BindAdjustments(modAdjustments = new AudioAdjustments());

    foreach (var mod in applicableToTrack)
        mod.ApplyToTrack(modAdjustments);
}
```

Note that this *ensures* that mods cannot ever do anything bad with the `AudioAdjustments` they're given, because the references they're given are thrown away after each `ResetTrackAdjustments()` call.